File: python/src/frtb/plat.py
"""P&L attribution test (PLAT): Spearman + KS metrics, traffic-light zone,
Amber capital surcharge.

Pinned thresholds:
  Green:  spearman >= 0.85  AND  KS <= 0.09
  Red:    spearman <  0.80  OR   KS >  0.12
  Amber:  everything in between.
Constant P&L on either side leaves the Spearman correlation undefined; the
desk is assigned RED in that case (documented conservative convention).

Amber surcharge (pinned k = 0.5 interpolation between IMA and SA):
  surcharge = k * max(0, SA_desk - IMA_desk_core)
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional, Sequence

from .params import SbmParams
from .stats import ks_statistic, spearman


@dataclass(frozen=True)
class PlatResult:
    """PLAT outcome; metrics are None when undefined (constant series -> Red)."""

    spearman: Optional[float]
    ks: Optional[float]
    zone: str  # 'green' | 'amber' | 'red'


def plat_zone_from_metrics(spearman_rho: float, ks_stat: float, params: SbmParams) -> str:
    """Map (Spearman, KS) to a PLAT zone using the pinned thresholds."""
    if not (math.isfinite(spearman_rho) and math.isfinite(ks_stat)):
        raise ValueError("plat_zone_from_metrics: metrics must be finite")
    if spearman_rho < params.plat_spearman_amber or ks_stat > params.plat_ks_amber:
        return "red"
    if spearman_rho >= params.plat_spearman_green and ks_stat <= params.plat_ks_green:
        return "green"
    return "amber"


def plat_test(hypo: Sequence[float], rtpl: Sequence[float], params: SbmParams) -> PlatResult:
    """Run the PLAT on hypothetical vs risk-theoretical P&L.

    A constant series on either side makes the rank correlation undefined:
    the result is Red with metrics = None (documented edge case).
    """
    if len(hypo) != len(rtpl):
        raise ValueError(f"plat_test: series length mismatch ({len(hypo)} vs {len(rtpl)})")
    if len(hypo) < 3:
        raise ValueError("plat_test: need at least 3 observations")
    try:
        rho = spearman(hypo, rtpl)
    except ValueError:
        # constant series -> correlation undefined -> Red (conservative)
        return PlatResult(spearman=None, ks=None, zone="red")
    ks = ks_statistic(hypo, rtpl)
    return PlatResult(spearman=rho, ks=ks, zone=plat_zone_from_metrics(rho, ks, params))


def plat_surcharge(zone: str, sa_capital: float, ima_capital_core: float,
                   params: SbmParams) -> float:
    """Amber-zone capital surcharge: k * max(0, SA - IMA_core); 0 otherwise.

    Red-zone desks fall back to SA entirely (handled by the caller/report);
    the surcharge formula itself applies only to Amber.
    """
    if zone not in ("green", "amber", "red"):
        raise ValueError(f"plat_surcharge: unknown zone '{zone}'")
    for name, v in (("sa_capital", sa_capital), ("ima_capital_core", ima_capital_core)):
        if not math.isfinite(v) or v < 0.0:
            raise ValueError(f"plat_surcharge: {name} must be >= 0 and finite")
    if zone != "amber":
        return 0.0
    return params.plat_k_surcharge * max(0.0, sa_capital - ima_capital_core)

File: java/src/main/java/com/quant/frtb/Plat.java
package com.quant.frtb;

/**
 * P&amp;L attribution test (PLAT): Spearman + KS metrics, traffic-light zone,
 * Amber capital surcharge.
 *
 * <p>Pinned thresholds: Green iff {@code spearman >= 0.85 AND KS <= 0.09};
 * Red iff {@code spearman < 0.80 OR KS > 0.12}; Amber otherwise. Constant
 * P&amp;L on either side leaves the rank correlation undefined; the desk is
 * assigned RED with null metrics (documented conservative convention).
 *
 * <p>Amber surcharge (pinned k = 0.5 interpolation between IMA and SA):
 * {@code surcharge = k * max(0, SA_desk - IMA_desk_core)}.
 */
public final class Plat {

    private Plat() {
    }

    /**
     * PLAT outcome; metrics are {@code null} when undefined (constant series,
     * which maps to Red).
     *
     * @param spearman Spearman rank correlation, or null when undefined
     * @param ks       two-sample KS statistic, or null when undefined
     * @param zone     "green" | "amber" | "red"
     */
    public record PlatResult(Double spearman, Double ks, String zone) {
    }

    /** Map (Spearman, KS) to a PLAT zone using the pinned thresholds. */
    public static String platZoneFromMetrics(double spearmanRho, double ksStat,
                                             SbmParams params) {
        if (!(Double.isFinite(spearmanRho) && Double.isFinite(ksStat))) {
            throw new IllegalArgumentException("platZoneFromMetrics: metrics must be finite");
        }
        if (spearmanRho < params.platSpearmanAmber() || ksStat > params.platKsAmber()) {
            return "red";
        }
        if (spearmanRho >= params.platSpearmanGreen() && ksStat <= params.platKsGreen()) {
            return "green";
        }
        return "amber";
    }

    /**
     * Run the PLAT on hypothetical vs risk-theoretical P&amp;L.
     *
     * <p>A constant series on either side makes the rank correlation
     * undefined: the result is Red with null metrics (documented edge case).
     */
    public static PlatResult platTest(double[] hypo, double[] rtpl, SbmParams params) {
        if (hypo.length != rtpl.length) {
            throw new IllegalArgumentException("platTest: series length mismatch ("
                    + hypo.length + " vs " + rtpl.length + ")");
        }
        if (hypo.length < 3) {
            throw new IllegalArgumentException("platTest: need at least 3 observations");
        }
        double rho;
        try {
            rho = Stats.spearman(hypo, rtpl);
        } catch (IllegalArgumentException e) {
            // constant series -> correlation undefined -> Red (conservative)
            return new PlatResult(null, null, "red");
        }
        double ks = Stats.ksStatistic(hypo, rtpl);
        return new PlatResult(rho, ks, platZoneFromMetrics(rho, ks, params));
    }

    /**
     * Amber-zone capital surcharge: {@code k * max(0, SA - IMA_core)}; 0 for
     * green/red (red-zone desks fall back to SA entirely — a reporting
     * matter handled by the caller, not a surcharge).
     */
    public static double platSurcharge(String zone, double saCapital, double imaCapitalCore,
                                       SbmParams params) {
        if (!"green".equals(zone) && !"amber".equals(zone) && !"red".equals(zone)) {
            throw new IllegalArgumentException("platSurcharge: unknown zone '" + zone + "'");
        }
        checkNonNegative("saCapital", saCapital);
        checkNonNegative("imaCapitalCore", imaCapitalCore);
        if (!"amber".equals(zone)) {
            return 0.0;
        }
        return params.platKSurcharge() * Math.max(0.0, saCapital - imaCapitalCore);
    }

    private static void checkNonNegative(String name, double v) {
        if (!Double.isFinite(v) || v < 0.0) {
            throw new IllegalArgumentException(
                    "platSurcharge: " + name + " must be >= 0 and finite");
        }
    }
}
